Add Volkswagen WeConnect integration

krzysdabro opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds integration for Volkswagen WeConnect service using [WeConnect-python](https://github.com/tillsteinbach/WeConnect-python) library.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#35099

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

krzysdabro committed:
Add Volkswagen WeConnect integration

autinerd commented on homeassistant/components/weconnect/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+from .coordinator import WeConnectCoordinator
+
+PLATFORMS = [
+    Platform.SENSOR,
+]
```
```suggestion
PLATFORMS = [Platform.SENSOR]
```

autinerd commented on homeassistant/components/weconnect/__init__.py:
```diff
@@ -0,0 +1,37 @@
+]
+
+
+@dataclass
+class WeConnectData:
+    """Class to store WeConnect runtime data."""
+
+    coordinator: WeConnectCoordinator
+
+
+type WeConnectConfigEntry = ConfigEntry[WeConnectData]
```
```suggestion
type WeConnectConfigEntry = ConfigEntry[WeConnectCoordinator]
```

No dataclass needed when there is only one object.

autinerd commented on homeassistant/components/weconnect/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    coordinator = WeConnectCoordinator(hass, config_entry)
+    await coordinator.async_config_entry_first_refresh()
+
+    config_entry.runtime_data = WeConnectData(coordinator)
```
```suggestion
    config_entry.runtime_data = coordinator
```

autinerd commented on homeassistant/components/weconnect/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    config_entry.runtime_data = WeConnectData(coordinator)
+
+    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
+    return True
```
```suggestion
    return True


async def async_unload_entry(
    hass: HomeAssistant, entry: WeConnectConfigEntry
) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```

autinerd commented on homeassistant/components/weconnect/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
```
```suggestion
    config_entry: WeConnectConfigEntry,
```

autinerd commented on homeassistant/components/weconnect/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+class WeConnectConfigFlow(ConfigFlow, domain=DOMAIN):
+    """WeConnect config flow."""
+
+    VERSION = 1
+
```
```suggestion
```

Not needed in the first version.

autinerd reviewed (changes requested):
Hi, thanks for the contribution! Here are some first comments.

EDIT: Can you please add a link to the library in the PR description?

krzysdabro committed:
Resolve issues

krzysdabro commented:
@autinerd Thank you for the review. I have fix the issues and included link to library in description.

joostlek reviewed (changes requested):
This library is scraping the web and that is not allowed via ADR-004. The exception states that every field should be queried at the same time, thus that means we should not use regex. Have a check at https://github.com/joostlek/python-volkswagen, we use the build in HTML parser and that is allowed in Home Assistant

krzysdabro commented:
@joostlek As far as I know, this library actually uses the JSON API for the data updates, as you can see [here](https://github.com/tillsteinbach/WeConnect-python/blob/main/weconnect/weconnect.py#L204-L205) and [here](https://github.com/tillsteinbach/WeConnect-python/blob/main/weconnect/elements/vehicle.py#L340-L341). he only time it interacts with the web application is during the authentication process, which, from my understanding, is allowed under ADR-004. While I must admit your library offers a much cleaner solution for the authentication process, it doesn’t provide further data on the vehicles.

What would you suggest as the solution here?

joostlek commented:
I know, I am also not saying that my library is the way to go, its just that authentication like that would be accepted, but auth like in this PR would not, since it uses regex parsing

krzysdabro commented:
FYI I have created a PR to refactor auth in the lib: https://github.com/tillsteinbach/WeConnect-python/pull/223

krzysdabro committed:
Fix coordinator reference

krzysdabro committed:
Update config flow

krzysdabro committed:
Fix brand mapping

krzysdabro committed:
Add quality scale

krzysdabro committed:
Update weconnect to 0.60.8

krzysdabro committed:
Update tests

krzysdabro committed:
Update translations

krzysdabro commented:
@joostlek The authentication process in the library has been refactored to use HTML parsing. I've upgraded the package, updated the translations, and added a quality scale checklist. Could you take a look when you have some spare time?
